File: reproduce_regex.js

const htmlContent = `
<p>Here is some text with a broken placeholder.</p>
<span class="variable">[Start of placeholder</span>
<p>Some intervening HTML content <span style="color: red">inside</span>.</p>
<span class="variable">End of placeholder]</span>
<p>Another valid placeholder [VALID_ONE]</p>
`;

const dangerousRegex = /\[([^\]]+)\]/g;
const safeRegex = /\[([^\]<>]+)\]/g;

console.log("Testing Dangerous Regex:");
let match;
while ((match = dangerousRegex.exec(htmlContent)) !== null) {
    console.log(`Found match: "${match[1]}"`);
}

console.log("\nTesting Safe Regex:");
while ((match = safeRegex.exec(htmlContent)) !== null) {
    console.log(`Found match: "${match[1]}"`);
}

File: reproduce_regex.py

import re

html_content = """
<p>Here is some text with a broken placeholder.</p>
<span class="variable">[Start of placeholder</span>
<p>Some intervening HTML content <span style="color: red">inside</span>.</p>
<span class="variable">End of placeholder]</span>
<p>Another valid placeholder [VALID_ONE]</p>
"""

dangerous_regex = r"\[([^\]]+)\]"
safe_regex = r"\[([^\]<>]+)\]"

print("Testing Dangerous Regex:")
matches = re.findall(dangerous_regex, html_content)
for match in matches:
    print(f"Found match: '{match}'")

print("\nTesting Safe Regex:")
matches = re.findall(safe_regex, html_content)
for match in matches:
    print(f"Found match: '{match}'")
